Feat: Log dumper

waldgange opened the pull request:
Introduce a set of parameters for log record dumping by a trigger message. This section can be added to the Broker config:
```
{
    ....
    "taskConfig": {
        ...
        "logDump": {
            "recordBufferSize": 32768,
            "recordingLevel": "TRACE",
            "triggerLevel": "FATAL"
        }
        ....
    }
    ....
}
```
Where:
`recordBufferSize` is a size of *default* logger's record buffer in bytes. 
`recordingLevel` is the lowest severity level of the record that will be stored by the logger in its log record buffer.
`triggerLevel` is the lowest severity level of the record that will cause immediate publication of that record and any records in the logger's log record buffer.

chrisbeard commented:
Looks like there's a unit test for the log controller, let's extend that to test this as well.

I think the most important thing to make sure of with this test is that we don't break compatibility with older configs that don't utilize this.

waldgange commented:
@chrisbeard I've added a unit test for `LogControllerConfig::fromObj()`, also I've removed the outdated method `fromDatum()` and its unit test.
In order not to break compatibility with old broker configs we have to release this feature in two steps. As the first step we release the new config, containing the new parameters. When all machines upgrade to this version we will release the broker that retrieves new parameters from the config

waldgange commented on src/groups/bmq/bmqtsk/bmqtsk_logcontroller.h:
```diff
@@ -512,7 +527,10 @@ int LogControllerConfig::fromObj(bsl::ostream& errorDescription,
-        .setSyslogFormat(obj.syslog().logFormat());
+        .setSyslogFormat(obj.syslog().logFormat())
+        .setRecordBufferSize(32768);
+    // TODO: use obj.logDump() when the config is updated
+    // .setRecordBufferSize(obj.logDump().recordBufferSize());
```
Ok I will remove this. This comment is a reminder to make the second step when all instances get the new version of the config. We used to leave such reminders in the past. Should we avoid them now?

waldgange commented on src/groups/bmq/bmqtsk/bmqtsk_logcontroller.h:
```diff
@@ -283,16 +298,6 @@ class LogControllerConfig {
-    /// might be partially altered.  Refer to the component level
-    /// documentation (section: "LogControllerConfig: Datum format") for the
-    /// expected format of the `datum`.
-    int fromDatum(bsl::ostream& errorDescription, const bdld::Datum& datum);
```
We don't use it anywhere since 2019. We use 'fromObj()' method to parse the broker config instead

waldgange commented on src/groups/bmq/bmqtsk/bmqtsk_logcontroller.cpp:
```diff
@@ -147,6 +147,9 @@ LogControllerConfig::LogControllerConfig(bslma::Allocator* allocator)
 , d_syslogAppName("")
 , d_syslogVerbosity(ball::Severity::ERROR)
 , d_categories(allocator)
+, d_recordBufferSize(32768)
```
This is the default `recordBufferSize` of the `ball::LoggerManagerConfiguration` component we use to dump logs. We used to set it here
https://github.com/bloomberg/blazingmq/blob/b39062018c2ef3e6b793aef8fb0e217401bbc642/src/groups/bmq/bmqtsk/bmqtsk_logcontroller.cpp#L654
Now it's a variable. We are going to retrieve this number from the config

hallfox replied:
Is this number in bytes? Can we just quickly mark the unit for it?

waldgange replied:
Yes. I've renamed the variable and added a comment in the same style as the variables above.

waldgange commented on src/groups/bmq/bmqtsk/bmqtsk_logcontroller.h:
```diff
@@ -522,6 +540,24 @@ int LogControllerConfig::fromObj(bsl::ostream& errorDescription,
+    //     errorDescription << "Invalid value for 'triggerLevel' ('"
+    //                      << obj.logDump().triggerLevel() << "')";
+    //     return -1;  // RETURN
+    // }
```
done

hallfox commented on src/groups/mqb/mqbcfg/mqbcfg_messages.cpp:
```diff
@@ -5967,6 +6131,13 @@ Configuration::print(bsl::ostream& stream, int level, int spacesPerLevel) const
+// GENERATED BY BLP_BAS_CODEGEN_2024.10.17
 // USING bas_codegen.pl -m msg --noAggregateConversion --noExternalization
 // --noIdent --package mqbcfg --msgComponent messages mqbcfg.xsd
+// ----------------------------------------------------------------------------
+// NOTICE:
+//      Copyright 2024 Bloomberg Finance L.P. All rights reserved.
+//      Property of Bloomberg Finance L.P. (BFLP)
+//      This software is made available solely pursuant to the
+//      terms of a BFLP license agreement which governs its use.
+// ------------------------------- END-OF-FILE --------------------------------
```
```suggestion
```
